### ISSUE TITLE
Spring Kafka Transaction - Rollback unsuccessful

### ISSUE THREAD

**Issue by agrawal93** [spring-projects/spring-kafka#455]
I have implemented a simple kafka transaction using `KafkaTransactionManager`, in which I send 3 messages to 2 different topics (`test_1`, `test_2`).
To test if transaction is working properly, I deleted topic `test_2` and I received `TopicNotFoundException` (which extends `RuntimeException`). My expectation is that messages don't get published to topic `test_1` too.

When I look at `TRACE` logs of `org.springframework.kafka` and `org.springframework.transaction` (see below), it says "Initiating transaction rollback". But, by that time, the consumer has already consumed from topic `test_1` (whichever messages were published to it).

I would like to know whether I am doing something wrong or I misunderstood the concept.

```log
2017-10-10 09:36:00.952 DEBUG 18000 --- [nio-8081-exec-9] o.s.k.t.KafkaTransactionManager          : Creating new transaction with name [agrawal93.kafka.test.services.Service.process]: PROPAGATION_REQUIRED,ISOLATION_DEFAULT; ''
2017-10-10 09:36:00.952 TRACE 18000 --- [nio-8081-exec-9] .s.t.s.TransactionSynchronizationManager : Bound value [org.springframework.kafka.core.KafkaResourceHolder@68e56fef] for key [org.springframework.kafka.core.DefaultKafkaProducerFactory@19f7222e] to thread [http-nio-8081-exec-9]
2017-10-10 09:36:00.952 DEBUG 18000 --- [nio-8081-exec-9] o.s.k.t.KafkaTransactionManager          : Created Kafka transaction on producer [CloseSafeProducer [delegate=org.apache.kafka.clients.producer.KafkaProducer@1dce2a97]]
2017-10-10 09:36:00.952 TRACE 18000 --- [nio-8081-exec-9] .s.t.s.TransactionSynchronizationManager : Initializing transaction synchronization
2017-10-10 09:36:00.952 TRACE 18000 --- [nio-8081-exec-9] o.s.t.i.TransactionInterceptor           : Getting transaction for [agrawal93.kafka.test.services.Service.process]
2017-10-10 09:36:02.866 TRACE 18000 --- [nio-8081-exec-9] .s.t.s.TransactionSynchronizationManager : Retrieved value [org.springframework.kafka.core.KafkaResourceHolder@68e56fef] for key [org.springframework.kafka.core.DefaultKafkaProducerFactory@19f7222e] bound to thread [http-nio-8081-exec-9]
2017-10-10 09:36:02.866 TRACE 18000 --- [nio-8081-exec-9] o.s.kafka.core.KafkaTemplate             : Sending: ProducerRecord(topic=test_1, partition=null, headers=RecordHeaders(headers = [], isReadOnly = false), key={"tenantId": "TEST"}, value={<some-value>}, timestamp=null)
2017-10-10 09:36:03.178 TRACE 18000 --- [nio-8081-exec-9] o.s.kafka.core.KafkaTemplate             : Sent: ProducerRecord(topic=test_1, partition=null, headers=RecordHeaders(headers = [], isReadOnly = true), key={"tenantId": "TEST"}, value={<some-value>}, timestamp=null)
2017-10-10 09:36:03.716 TRACE 18000 --- [nio-8081-exec-9] o.s.t.i.TransactionInterceptor           : Completing transaction for [agrawal93.kafka.test.services.Service.process] after exception: agrawal93.kafka.test.exceptions.TopicNotFoundException: Unable to publish to Kafka. Topic [test_2] does not exist.
2017-10-10 09:36:03.716 TRACE 18000 --- [nio-8081-exec-9] o.s.t.i.RuleBasedTransactionAttribute    : Applying rules to determine whether transaction should rollback on agrawal93.kafka.test.exceptions.TopicNotFoundException: Unable to publish to Kafka. Topic [test_2] does not exist.
2017-10-10 09:36:03.716 TRACE 18000 --- [nio-8081-exec-9] o.s.t.i.RuleBasedTransactionAttribute    : Winning rollback rule is: null
2017-10-10 09:36:03.716 TRACE 18000 --- [nio-8081-exec-9] o.s.t.i.RuleBasedTransactionAttribute    : No relevant rollback rule found: applying default rules
2017-10-10 09:36:03.716 TRACE 18000 --- [nio-8081-exec-9] o.s.k.t.KafkaTransactionManager          : Triggering beforeCompletion synchronization
2017-10-10 09:36:03.716 DEBUG 18000 --- [nio-8081-exec-9] o.s.k.t.KafkaTransactionManager          : Initiating transaction rollback
2017-10-10 09:36:03.878 TRACE 18000 --- [nio-8081-exec-9] .s.t.s.TransactionSynchronizationManager : Clearing transaction synchronization
2017-10-10 09:36:03.878 TRACE 18000 --- [nio-8081-exec-9] o.s.k.t.KafkaTransactionManager          : Triggering afterCompletion synchronization
2017-10-10 09:36:03.878 TRACE 18000 --- [nio-8081-exec-9] .s.t.s.TransactionSynchronizationManager : Removed value [org.springframework.kafka.core.KafkaResourceHolder@68e56fef] for key [org.springframework.kafka.core.DefaultKafkaProducerFactory@19f7222e] from thread [http-nio-8081-exec-9]
agrawal93.kafka.test.exceptions.TopicNotFoundException: Unable to publish to Kafka. Topic [test_2] does not exist.
```

Kindly advice. Thanks.

**Comment by garyrussell** [comment]
Please show your code and configuration. You should also enable debug/trace logging for `org.apache.kafka` and include the full log.

**Comment by artembilan** [comment]
@agrawal93 ,

Thank you for sharing that, but it's hard to follow too much code here in the narrow GH window.
Any chances to make a simple Spring Boot app and point to the repo on GH?
We will fork it and play locally.

**Comment by agrawal93** [comment]
@artembilan 
Apologies for the inconvenience. Please find my `kafka-txn-test` github repository [here](https://github.com/agrawal93/kafka-txn-test)

Pre-requisites for running the code:
1. Kafka 0.11.0.0 --- Need to be configured [kafka.properties](https://github.com/agrawal93/kafka-txn-test/blob/master/src/main/resources/kafka.properties)
2. Confluent Schema Registry --- Need to be configured [kafka.properties](https://github.com/agrawal93/kafka-txn-test/blob/master/src/main/resources/kafka.properties)

Once deployed, place a get request on `http://localhost:8080/txns/` to start the process.

**Comment by artembilan** [comment]
Sorry, your application is very complicated. 
I don't know what is "Confluent Schema Registry" and I'm sure it is fully unrelated to the problem.

What I expect from you is something like simple transactional producer and the fact that messages are consumed. That should be something to easy run as a test case...

**Comment by agrawal93** [comment]
Oh okay. Then I think it must be an issue with either Kafka or Confluent.
Well, thanks for the help.

**Comment by l308718280** [comment]
Have you resolved this problem?

**Comment by agrawal93** [comment]
Yes. Apparently, it was the issue with Confluent-provided Avro Serializers and Deserializers. I was using the old version of `kafka-avro-serializer`, i.e. `3.2.2`. Updating the version `3.3.0` worked for me.
